File: 64.最小路径和/minPathSum.go
package problem0064

// 同62题，不同的是，需要找出一条从左上到右下的路径，使得路径上的数字总和为最小
// 步骤一：定义数组元素的含义 => 目的是最小路径和，那么dp[i][j]代表从左上角走到(i, j)时，最小的路径和是dp[i][j]
// 步骤二：找出数组元素之间的关系 => 显然，dp[i][j] = min(dp[i][j-1], dp[i-1][j]) + arr[i][j]，即，当前位置的值 + 前一步两种可能的最小值
// 步骤三：找出初始值 => 同62题
//        1. 对于第一行：dp[0][j] = dp[0][j-1] + arr[0][j]
//        2. 对于第一列：dp[j][0] = dp[j-1][0] + arr[j][0]
func minPathSum(grid [][]int) int {
	for i := 1; i < len(grid); i++ {
		grid[i][0] += grid[i-1][0]
	}
	for j := 1; j < len(grid[0]); j++ {
		grid[0][j] += grid[0][j-1]
	}
	for i := 1; i < len(grid); i++ {
		for j := 1; j < len(grid[0]); j++ {
			grid[i][j] += min(grid[i-1][j], grid[i][j-1])
		}
	}
	return grid[len(grid)-1][len(grid[0])-1]
}

func min(a, b int) int {
	if a > b {
		return b
	}
	return a
}

File: 64.最小路径和/minPathSum.js
// 给定一个包含非负整数的 m x n 网格 grid ，请找出一条从左上角到右下角的路径，使得路径上的数字总和为最小。
// 每次只能向下或者向右移动一步。

// 状态：dp[i][j]代表完成(i+1) * (j+1)的网格的最小路径和
// 状态转移方程：dp[i][j] = Min(dp[i][j-1], dp[i-1][j]) + grid[i][j]
// 初始值：
//   对于第一行：dp[0][j] = dp[0][j-1] + grid[0][j]
//   对于第一列：dp[i][0] = dp[i-1][0] + grid[i][0]
//   dp[0][0] = grid[0][0]
// 结果：dp[grid.length-1][grid[0].length-1]
function minPathSum(grid) {
    const dp = Array.from({ length: grid.length }, () => Array.from({ length: grid[0].length }, () => Number.MAX_VALUE))
    dp[0][0] = grid[0][0]
    // 第一列
    for (let i = 1; i < grid.length; i++) {
        dp[i][0] = dp[i-1][0] + grid[i][0]
    }
    // 第一行
    for (let j = 1; j < grid[0].length; j++) {
        dp[0][j] = dp[0][j-1] + grid[0][j]
    }
    for (let i = 1; i < grid.length; i++) {
        for (let j = 1; j < grid[0].length; j++) {
            dp[i][j] = Math.min(dp[i][j-1], dp[i-1][j]) + grid[i][j]
        }
    }
    return dp[grid.length-1][grid[0].length-1]
}

// 状态压缩（投影）技巧：将dp[i][j]的状态数组压缩为dp[j]，空间复杂度从 O(N^2) 降低到 O(N)。
// 特征：状态转移方程需要的都是dp[i][j]相邻的状态
// 本题中需要dp[i][j-1]和dp[i-1][j]两个相邻状态
// (i-1, j-1) (i, j-1)
// (i, j-1)   (i, j)
// 在遍历过程中，
function minPathSum(grid) {
    const dp = Array.from({ length: grid.length }, () => Number.MAX_VALUE)
    dp[0] = grid[0][0]
    // 第一行
    for (let j = 1; j < grid[0].length; j++) {
        dp[j] = dp[j-1] + grid[0][j]
    }
    for (let i = 1; i < grid.length; i++) {
        dp[0] += grid[i][0]
        for (let j = 1; j < grid[0].length; j++) {
            dp[j] = Math.min(dp[j-1], dp[j]) + grid[i][j]
        }
    }
    return dp[grid[0].length-1]
}